src/main/java/co/cask/gcp/bigquery/BigQueryUtils.java

```diff
@@ -0,0 +1,247 @@
+
+    // Complex types like arrays, maps and unions are not supported in BigQuery plugins.
+    if (!type.isSimpleType()) {
+      throw new IllegalArgumentException(String.format("Field '%s' is of unsupported type '%s'.", name, type));
+    }
+  }
+
+  /**
+   * Gets non nullable type from provided schema.
+   *
+   * @param schema schema to be used
+   * @return non-nullable {@link Schema}
+   */
+  static Schema getNonNullableSchema(Schema schema) {
+    return schema.isNullable() ? schema.getNonNullable() : schema;
+  }
+
+  /**
+   * Get Bigquery {@link Configuration}.
+   *
+   * @param serviceAccountFilePath service account file path
+   * @param projectId BigQuery project ID
+   * @return {@link Configuration} with config set for BigQuery
+   * @throws IOException if not able to get credentials
+   */
+  static Configuration getBigQueryConfig(@Nullable String serviceAccountFilePath, String projectId) throws IOException {
+    Job job = Job.getInstance();
+
+    // some input formats require the credentials to be present in the job. We don't know for
+    // sure which ones (HCatalog is one of them), so we simply always add them. This has no other
+    // effect, because this method is only used at configure time and will be ignored later on.
+    if (UserGroupInformation.isSecurityEnabled()) {
+      Credentials credentials = UserGroupInformation.getCurrentUser().getCredentials();
+      job.getCredentials().addAll(credentials);
+    }
+
+    Configuration configuration = job.getConfiguration();
+    configuration.clear();
+    if (serviceAccountFilePath != null) {
+      configuration.set("mapred.bq.auth.service.account.json.keyfile", serviceAccountFilePath);
+      configuration.set("google.cloud.auth.service.account.json.keyfile", serviceAccountFilePath);
+    }
+    configuration.set("fs.gs.impl", "com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem");
+    configuration.set("fs.AbstractFileSystm.gs.impl", "com.google.cloud.hadoop.fs.gcs.GoogleHadoopFS");
+    configuration.set("fs.gs.project.id", projectId);
+    configuration.set(BigQueryConfiguration.PROJECT_ID_KEY, projectId);
+    return configuration;
+  }
+
+  /**
+   * Get BigQuery Table.
+   *
+   * @param serviceAccountFilePath service account file path
+   * @param project BigQuery project ID
+   * @param dataset dataset for the BigQuery table
+   * @param table BigQuery table
+   * @return returns BigQuery table
+   * @throws IOException if not able to load credentials
+   */
+  @Nullable
+  static Table getBigqueryTable(@Nullable String serviceAccountFilePath, String project,
+                                String dataset, String table) throws IOException {
+    BigQuery bigquery;
+    BigQueryOptions.Builder bigqueryBuilder = BigQueryOptions.newBuilder();
+    if (serviceAccountFilePath != null) {
+      bigqueryBuilder.setCredentials(loadServiceAccountCredentials(serviceAccountFilePath));
+    }
+    if (project == null) {
```

Comment:
project is not annotated as nullable. As long as plugins use the getProject() method, this should be handled by the logic in the plugin config.